### PR TITLE
support keepHistory (or --keep-history) flag

### LINKED ISSUE #35: add option to not delete commit history
I typically keep a ref to the skeleton repo around, because I'm constantly improving my own skeletons, and I often pull the improvements into projects based off of them.

### PR DESCRIPTION
**Pull request by jedwards1211**
if given, pollinate won't delete the .git directory

fix #35

### CODE REVIEW

File: lib/init.js [hunk 1]
```diff
@@ -14,7 +14,10 @@ var isFile = require('file-exists');
 var isDirectory = require('is-directory').sync;
 var isGitURL = require('is-git-url');
 var isGitHub = function (input) { return input.match(/^(?!\.)[A-Za-z0-9\-_]+\/[A-Za-z0-9\-_]+$/g); };
-var isGit = function (input) { return isGitURL(input) || isURL(input); };
+var isGit = function (input) {
+    return isGitURL(input) || isURL(input) ||
+        (input.startsWith('git@github.com:') && isGitHub(input.substring('git@github.com:'.length)));
```
Comment (jedwards1211) [inline lib/init.js]: @howardroark also note this change I made, maybe should have been in a separate PR

File: tests/test.js [hunk 2]
```diff
@@ -8,6 +8,12 @@ var fs = require('fs');
 var path = require('path');
 
 describe('Test basic example', function () {
+    afterEach(function (done) {
+        rimraf('newproject', function () {
+            rimraf('test', done);
+        });
+    });
+
```
Comment (jedwards1211) [inline tests/test.js:16]: I moved all `rimraf`ing into this `afterEach` hook

File: lib/init.js (hunk 1 above)
Comment (howardroark) [reply lib/init.js]: Could you maybe explain this a bit more... I'm a little confused :P

**Comment by howardroark** [comment]
Woohoo!  I'll try this out locally soon.

File: lib/fetch.js [hunk 3]
```diff
@@ -125,7 +125,7 @@ var fetchGit = function (state, callback) {
             }
         },
         function (next) {
-            if (isDirectory(tmpPath + '/.git')) {
+            if (isDirectory(tmpPath + '/.git') && !state.options.keepHistory && !(state.flags && state.flags['keep-history'])) {
```
Comment (howardroark) [inline lib/fetch.js:128]: Ahhh right... extend has not occurred yet so `state.keepHistory` is not available.  So you couldn't keep this in `~/.pollen`. Darn.

Maybe just check for the flag here and remove the `options.keepHistory` part.  Considering both would come from the cli and have the same objective.

File: tests/test.js (hunk 2 above)
Comment (howardroark) [reply tests/test.js:16]: 👍

**Comment by jedwards1211** [comment]
I'm setting up my [es2015-library-skeleton](https://github.com/jedwards1211/es2015-library-skeleton/tree/pollinate) project to use `pollinate` now, check it out!

**Comment by howardroark** [comment]
Awesome!  Well I'm happy you are finding this useful.  It's great to have more feedback/ideas.

**Review by howardroark — APPROVED** [review]

**Comment by jedwards1211** [comment]
I completely forgot to document this in the README, I'll make another PR for that

**Comment by jedwards1211** [comment]
@howardroark also, I didn't realize until now, this doesn't actually make it possible for me to pull in any and all changes from the project skeleton, because I would be pulling in changes to the template files, rather than the results of pollinate transformation.  D'oh!

I'll have to think for a bit about how to deal with this.  What I'm thinking is a `pollinate pull` command that pulls changes from the skeleton into a new branch, does the usual transform, then merges back into the current branch.

**Comment by jedwards1211** [comment]
@howardroark for more background on this: I've sometimes made improvements to my project skeletons that I want to easily pull in to all of my projects that are using them.  For instance, I've added various helpful npm scripts, `semantic-release`, switched from `pre-commit` to `husky`, etc.

**Comment by howardroark** [comment]
@jedwards1211 Yeah... I was thinking you would just be using this to run diffs and maybe sync assets.

My view is that you benefit in new projects as your base evolves from learnings in existing ones. I think a certain amount of refactoring is always required in getting new things to work in existing projects...

**Comment by jedwards1211** [comment]
Okay, I restructured my skeleton projects in such a way that I can at least pull in `package.json` updates that don't involve template variables.  Instead of just having pollinate parse `package.json` directly or parsing a `PROJECT-package.json` and then moving it into `package.json`'s place, I put only template stuff in `PROJECT-package.json`:
```json
{
  "name": "{{ name }}",
  "version": "0.0.0-development",
  "description": "{{ description }}",
  "repository": {
    "type": "git",
    "url": "https://github.com/{{ organization }}/{{ name }}.git"
  },
  "author": "{{ author }}",
  "bugs": {
    "url": "https://github.com/{{ organization }}/{{ name }}/issues"
  },
  "homepage": "https://github.com/{{ organization }}/{{ name }}#readme"
}
```

And then I merge it with this script (run from `complete):
```js
#!/usr/bin/env node

var merge = require('lodash.merge')
var fs = require('fs')
var merged = merge(require('./package.json'), require('./PROJECT-package.json'))
fs.writeFileSync('./package.json', JSON.stringify(merged, null, 2), 'utf8')
```

That way the skeleton project can have a valid `package.json` that I can update, but its `name`, `description`, etc. get overwritten by the parsed fields.

So yeah, I think this will prove useful to me!

File: lib/fetch.js (hunk 3 above)
Comment (jedwards1211) [reply lib/fetch.js:128]: Oh, huh.  Well I didn't actually need it from `~/.pollen`, I was just trying to mimic everything else, so that's fine with me.

Comment (jedwards1211) [reply lib/fetch.js:128]: Wait a minute, it does have an effect on one of the tests I wrote for it: https://github.com/howardroark/pollinate/blob/develop/tests/test.js#L93.  Not via `~/.pollen` but via `options` passed to the node API.

**Comment by howardroark** [comment]
I wonder if a `merge` step would make sense?  Would allow you to discard a lot of this if it ran before.

**Comment by howardroark** [comment]
I kinda see two things going on here...

1. Add a `merge` after `parse` that takes and an array of json files to merge on top of each other
2. Move `discard` to after `move` (so that the `.git` removal and check can go there)

```
merge: [
  ["package.json", "PROJECT-package.json"]
]
```

Then you don't need that script.  Plus you can also then discard `PROJECT-pacakge.json`.  The git upstream won't care if you are just merging down updates and that commit is in the same repo.

**Comment by jedwards1211** [comment]
@howardroark That would be cool!  I wasn't sure if you would consider it worthy of being a specific feature in pollinate.  My only concern is whether people would want to merge other types of files (YAML, XML, etc), though it could operate based upon the extension.

**Comment by howardroark** [comment]
I'd say start with JSON and and if someone wants to add other object file formats... not a big deal to support.  I think it's nice to have example configs that don't have all sort of template tags!

**Pull request merged**